keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,560 @@
+- [Validates Job cleanup](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller_test.go#L371) path of the controller under different conditions.
+- [Validates Status](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller_test.go#L593) of the CronJob after sync under different conditions.
+
+## Implementation plan
+
+#### Release X
+
+- Alpha: Feature flag for new controller is disabled by default
+- Dual controller. Both old and new implementation co-exist
+
+#### Release X+1
+
+- Beta: Feature flag for new controller is enabled by default. If the distribution chooses it can be disabled
+- Introduce CronJob in batch/v1
+
+#### Release X+2
+
+- GA: The feature flag is deprecated and the old controller code is cleaned up
+
+## Graduation Criteria
+
+- [ ] Implement shared informers to reduce pressure on API Server
+- [ ] Pass conformance tests
+- [ ] Update documents reflecting the changes
+- [ ] Pass CronJob e2e tests
+- [ ] Pass CronJob unit-tests
+- [ ] Pass scale tests
+
+## Implementation History
+
+- CronJob was introduced in Kubernetes 1.3 as ScheduledJobs
+- In Kubernetes 1.8 it was renamed to CronJob and promoted to Beta
+- In Kubernetes 1.19 CronJob was marked deprecated with removal in 1.22 due to not being promoted to GA.
+
+## Alternatives and Further Reading
+
+#### Cron Aspect
+
+To detect which CronJob has met its schedule and need to create Jobs we need to implement a timer component. These are the possible options for implementing the timer:
+
+| algorithm  | how it works | notes |
+|:----------|:----------|:-------|
+| Unordered timer list | Periodic Sweep from cache | Slower and similar to existing implementation. But improved because we sweep fom the cache instead of API server |
+|Ordered timer list| Maintain ordered list of Cronjob keys and next time of expiry. Keep starting a timer with the earliest expiry. | Efficient. Reinsertion to list takes O(n) |
+|Timer trees| Instead of ordered list use a sorted binary tree. | More efficient. Insertion is O(log n) |
+|Heap based timer|A variant of ordered timer list where heap is used to store the next expiry time | Efficient compared to ordered list. Bookkeeping and insertion is O(log n). |
+|Simple Timing wheels| circular buffer of MaxTimeOut slots. List of expiring timers at each slot. | Works for small bounded  MaxTimeOut which is not our case. Insertion and removal is O(1) via indexing |
+|Hashed Wheel| Hash expiring time and insert in a hash table with linked list at each index | Bookkeeping is O(1) and worst case insertion is O(n) |
+|Hierarchical Wheel| multiple timer wheels for different resolutions (Seconds, minutes, hours, days). When seconds rolls over we grab the next minutes timers and recreate the seconds wheel. similarly for minutes and hours. | Sharding at different hierarchy levels improves insertion and bookkeeping performance. |
+
+## Improvements/Considerations
+
```

Comment:
The improvements listed below aren't blockers for CronJob GA, I presume?

Reply:
Added comment to make it clear.